Update for 1.25

starlm opened the pull request:
Should work on 1.20 and newer but I haven't tested anything below 1.24.
I've been using it for a week now and encountered no problems.

Arcs & chains support is untested as of right now but it should work.

Edit: Arcs & chains support is implemented in a way to not bring the average down, right now it just sets it to 1.0 if it's not being rated by the game

Commit:
Update for 1.25

Commit:
Bump dependencies

kinsi55 commented:
I will try to look into this some time soon thank you

starlm commented:
Just wanted to give a quick update on 1.20+ support. 
I recently finished an implementation for measuring overswing on arcs and chains reliably. 
Branch: [OrdinaryTable:arc-chain-support](https://github.com/OrdinaryTable/BeatSaber_OverswingCounter/tree/arc-chain-support)

In case you want them to be measured, I could merge it into [OrdinaryTable:1.25](https://github.com/OrdinaryTable/BeatSaber_OverswingCounter/tree/1.25).

I've been using it for some weeks now with no problems and no changes to the actual score, as that would of course be quite bad.

Note: Implementation could technically be simplified by a lot but I'm worried that it might end up changing the score in certain scenarios

kinsi55 commented:
Hey, thanks for the update and sorry that I didnt get around to looking into this yet. I think arcs and chains should just be ignored unless explicitly enabled (opt-in)

 I'll compile this PR and try it next time I play

kinsi55 commented:
Neat to see that it supposedly works in replays now? Did you validate that? Based of a quick look and knowing my left hand the values dont look completely off hence me asking :D

starlm commented:
It does work in replays but only in certain scenarios, the play style (i.e. swing speed, etc.) heavily affects the accuracy of the overswing. When comparing it to live data it's slightly off.

Commit:
Merge arc-chain-support into 1.25

Commit:
Added option to ignore arcs & chains

Commit:
logic error

starlm commented:
5f52d36 fixed arcs & chains not being rated in a certain scenario.

I would also suggest completely disabling the counter when watching a ScoreSaber/BeatLeader replay, as SS replays result in inaccurate data and BL replays don't provide any ratings at all (i.e. 0% pre and 0% post)

kinsi55 commented:
Probably just display - for the values in those cases yeah

Commit:
Increase reliability of UnlockRatings

Harmony sometimes got Brfalse when dnSpy got Brfalse.s

Commit:
Disable counter in replays

kinsi55 commented:
I was gonna go over all your transpilers but I'm starting to wonder if, instead of transpiling the game to hell and back, it wouldnt instead make sense to instead just seperately track the saber movements and calculate the angles manually on note cut events.

This way it would also properly work in replays, and be much less prone to breaking on updates

starlm commented:
It might be a better solution but we would need to do all of the checks which the game does ourselves, i.e. when did the saber reach the center of the note, when does the swing actually end, etc.

I highly doubt that BeatGames will do any changes to how cuts are calculated any time soon, as the only thing which actually broke in 1.20, even tho it changed quite a lot, was the way it created a new CutInfo instance.

But at the end it's really a question of how much more complicated or prone to breaking it would be.

kinsi55 commented:
> i.e. when did the saber reach the center of the note, when does the swing actually end, etc.

This actually isnt a thing - Angle calculation is extremely primitive. Preswing is the angle up until the point when you touch the note hitbox, and postswing is everything after

starlm commented:
I see. If we do it your way, it would probably be best to just completely rewrite it from scratch (or at least everything but the custom counter).

kinsi55 commented:
Yeah, that would make the most sense. It could reuse a lot of the basegame components that are used for angle calculation, just with adjusted logic / removed clamping (Methods can be generated at runtime with copied / adjusted IL)

I'll probably still test / release your current solution so that something does exist for now and leave this as something to worry about later :tm:

starlm commented:
Sounds good. I'll get started on the rewrite in the mean time, if that's okay with you.

kinsi55 commented:
If you wanna look into that sure, be my guest

starlm commented:
Just wanted to give a quick update on how the rewrite is going.
It's coming along very well and so far it has 0 transpilers and 1 prefix (I'm not expecting that to change).
By the way it works in ScoreSaber and BeatLeader replays, I have not tested the accuracy yet but it looks pretty good.

I'll give you another update once it's done (Presumingly sometime next week).

kinsi55 commented:
Thats about what I would have expected, sounds great!

starlm commented:
I looked into replay support and I don't think that accurate data will be possible. Due to beat saber adding new data to the ``SaberMovementData`` instance each frame, it simply won't be possible unless the frame rate matches the replay's frame rate.

That being said, it certainly is far more unlikely to break with updates and more accurate now but that's about it.

kinsi55 commented:
Even if its not accurate, it should be a good enough estimate I'd guess - All of this really doesnt need to be extremely accurate as its really only meant to give you an idea how close you are to underswinging / how much you are overswinging / what your balance looks like

starlm commented:
I should've worded it differently, sometimes it's a good 20-40° off. But just like before, it's heavily affected by the play style.
I'll try to make a comparison video later today to show how far off it is compared to live data.

kinsi55 commented:
Dont need to worry about that, I'll see for myself whenever I happen to test it 👍

starlm commented:
Rewrite is done. Overswing appears to be pretty accurate. In terms of replays, ScoreSaber performed way better than expected. BeatLeader on the other hand, performed horrendously bad.

I made a little side-by-side comparison video: https://youtu.be/V0UNAkmzXPU

If you want to try it yourself, I could send you the source code over Discord or create a new branch, whatever you prefer.

kinsi55 commented:
You could just commit it into this PR's branch

I wonder why it is so bad on Beatleader - I have a feeling its down to pulling the values at the wrong frame seeing as consistently the upper value is bigger by about the amount missing at the bottom.

Commit:
Rewrite

kinsi55 committed:
misc

kinsi55 committed:
PluginConfig => Config

kinsi55 committed:
tabs

kinsi55 committed:
Less runtime allocation

kinsi55 committed:
Need to unsubscribe from the event too

kinsi55 committed:
Minimize accessor calls

kinsi55 committed:
names

kinsi55 committed:
todo

Commit:
cleanup

starlm commented on Models/DataProcessor.cs:
```diff
@@ -50,6 +50,8 @@ public class DataProcessor : ISaberMovementDataProcessor {
 				var afterCutTopPos = newData.topPos;
 				var afterCutBottomPos = newData.bottomPos;
+
+				// Raycasts are somewhat expensive - Might make sense to reuse the SSRC's values with a transpiler
```
I looked into it and it tended to be 3-5x slower. When reusing the values, it sometimes didn't get them for some reason but it reused 85-90% of SSRC's values.
It's mostly slower due to it making sure that the newest ``BladeMovementDataElement`` is the same to not mess up overswing.

kinsi55 replied:
Weird - I guess your transpiler just stored the result in some static field that you then retrieved?

starlm replied:
It added the ray result into a dictionary (`Dictionary<SaberSwingRatingCounter, Dictionary<BladeMovementDataElement, Vector3>>`), `DataProcessor` then tries to get the result from it. It sounds very inefficient, I know, but not using the SSRC and `BladeMovementDataElement` to get the correct one, completely messed up the overswing.

kinsi55 replied:
Shouldnt the singlethreaded nature mean that the basegame `ProcessNewData` is always executed before the counters one is, and thus you'd only need the "last" value?

starlm replied:
That's what I tried before, it tended to double the overswing.

kinsi55 replied:
Hm, well, realisticly it shouldnt really matter, was just a possible optimization that came to mind

kinsi55 committed:
avoid copying NoteCutInfo struct

Commit:
decrease code repetition

Commit:
request last data

Commit:
simplify data processor

kinsi55 commented:
[Played with it yesterday](https://www.twitch.tv/videos/1639708051). I didnt pay all that much attention to it but generally the numbers did not seem to be "obviously wrong". I'll try give it a more technical test some time soon

Commit:
more optimization

remove useless variable declaration
calculate swing angle after swing rating computation

kinsi55 commented:
Small update, really sorry I didnt merge / release this yet. I've had this plugin installed for all that time now its just that every time I play I forget to actually give it a deeper look :D I'll try to remember next time I play

kinsi55 commented:
Looked at it today briefly, seems correct at first glance. I'll be releasing this in a bit 👍
